testnet/stacks-node/src/nakamoto_node/miner.rs

```diff
@@ -0,0 +1,621 @@
+            self.keychain.generate_proof(
+                VRF_MOCK_MINER_KEY,
+                self.burn_block.sortition_hash.as_bytes(),
+            )
+        } else {
+            self.keychain.generate_proof(
+                self.registered_key.target_block_height,
+                self.burn_block.sortition_hash.as_bytes(),
+            )
+        };
+
+        debug!(
+            "Generated VRF Proof: {} over {} ({},{}) with key {}",
+            vrf_proof.to_hex(),
+            &self.burn_block.sortition_hash,
+            &self.burn_block.block_height,
+            &self.burn_block.burn_header_hash,
+            &self.registered_key.vrf_public_key.to_hex()
+        );
+        Some(vrf_proof)
+    }
+
+    /// Try to mine a Stacks block by assembling one from mempool transactions and sending a
+    /// burnchain block-commit transaction.  If we succeed, then return the assembled block data as
+    /// well as the microblock private key to use to produce microblocks.
+    /// Return None if we couldn't build a block for whatever reason.
+    fn mine_block(&mut self) -> Option<NakamotoBlock> {
+        debug!("block miner thread ID is {:?}", thread::current().id());
+        neon_node::fault_injection_long_tenure();
+
+        let burn_db_path = self.config.get_burn_db_file_path();
+
+        // NOTE: read-write access is needed in order to be able to query the recipient set.
+        // This is an artifact of the way the MARF is built (see #1449)
+        let mut burn_db =
+            SortitionDB::open(&burn_db_path, true, self.burnchain.pox_constants.clone())
+                .expect("FATAL: could not open sortition DB");
+
+        let mut chain_state = neon_node::open_chainstate_with_faults(&self.config)
+            .expect("FATAL: could not open chainstate DB");
+
+        let mut mem_pool = self
+            .config
+            .connect_mempool_db()
+            .expect("Database failure opening mempool");
+
+        let target_epoch_id =
+            SortitionDB::get_stacks_epoch(burn_db.conn(), self.burn_block.block_height + 1)
+                .ok()?
+                .expect("FATAL: no epoch defined")
+                .epoch_id;
+        let mut parent_block_info = self.load_block_parent_info(&mut burn_db, &mut chain_state)?;
+        let vrf_proof = self.make_vrf_proof()?;
+
+        if self.last_mined_blocks.is_empty() {
+            if parent_block_info.parent_tenure.is_none() {
+                warn!(
+                    "Miner should be starting a new tenure, but failed to load parent tenure info"
+                );
+                return None;
+            }
+        }
+
+        // create our coinbase if this is the first block we've mined this tenure
+        let tenure_start_info = if let Some(ref par_tenure_info) = parent_block_info.parent_tenure {
+            let parent_block_id = parent_block_info.stacks_parent_header.index_block_hash();
+            let current_miner_nonce = parent_block_info.coinbase_nonce;
```

Comment:
Would it make more sense / be more robust to just query the miner's account here to get the nonce?

Reply:
That is what is queried in `MinerThread::load_block_parent_info` to get that field.

Reply:
Oh I see, you're doing this on the call to `load_block_parent_info()`, right?

Reply:
Yep, exactly